File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/execution/CountEvaluator.java
package org.opencds.cqf.cql.engine.elm.execution;

import java.util.Iterator;

import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument;
import org.opencds.cqf.cql.engine.execution.Context;

/*
Count(argument List<T>) Integer

* The Count operator returns the number of non-null elements in the source.
* If the list contains no non-null elements, the result is 0.
* If the list is null, the result is null.
* Always returns Integer
*/

public class CountEvaluator extends org.cqframework.cql.elm.execution.Count {

    public static Object count(Object source) {
        if (source == null) {
            return 0;
        }

        Integer size = 0;

        if (source instanceof Iterable) {
            Iterable<?> element = (Iterable<?>)source;
            Iterator<?> itr = element.iterator();

            if (!itr.hasNext()) { // empty list
                return size;
            }

            while (itr.hasNext()) {
                Object value = itr.next();

                if (value == null) { // skip null
                    continue;
                }

                ++size;
            }

            return size;
        }

        throw new InvalidOperatorArgument("Count(List<T>)", String.format("Count(%s)", source.getClass().getName()));
    }

    @Override
    protected Object internalEvaluate(Context context) {
        Object source = getSource().evaluate(context);
        return count(source);
    }
}

File: Src/java/engine/src/commonMain/kotlin/org/opencds/cqf/cql/engine/elm/executing/CountEvaluator.kt
package org.opencds.cqf.cql.engine.elm.executing

import kotlin.jvm.JvmStatic
import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument
import org.opencds.cqf.cql.engine.runtime.Integer
import org.opencds.cqf.cql.engine.runtime.List
import org.opencds.cqf.cql.engine.runtime.Value
import org.opencds.cqf.cql.engine.runtime.toCqlInteger

/*
Count(argument List<T>) Integer

* The Count operator returns the number of non-null elements in the source.
* If the list contains no non-null elements, the result is 0.
* If the list is null, the result is null.
* Always returns Integer
*/
object CountEvaluator {
    @JvmStatic
    fun count(source: Value?): Integer {
        if (source == null) {
            return Integer(0)
        }

        var size = 0

        if (source is List) {
            val element = source
            val itr = element.iterator()

            if (!itr.hasNext()) { // empty list
                return size.toCqlInteger()
            }

            while (itr.hasNext()) {
                val value = itr.next()

                if (value == null) { // skip null
                    continue
                }

                ++size
            }

            return size.toCqlInteger()
        }

        throw InvalidOperatorArgument("Count(List<T>)", "Count(${source.typeAsString})")
    }
}
